File: packages/scala/oauth/src/main/scala/do/oauth/DeviceFlow.scala
package `do`.oauth

import sttp.client3._
import sttp.client3.circe._
import io.circe._
import io.circe.generic.auto._
import io.circe.parser._
import scala.util.{Try, Success, Failure}

case class DeviceAuthResponse(
  deviceCode: String,
  userCode: String,
  verificationUri: String,
  expiresIn: Int,
  interval: Int
)

case class TokenResponse(
  accessToken: String,
  refreshToken: Option[String],
  tokenType: String,
  expiresIn: Option[Int]
)

/**
 * Device flow authentication.
 */
class DeviceFlow(clientId: String) {
  private val backend = HttpURLConnectionBackend()
  private val authUrl = "https://auth.apis.do/user_management/authorize_device"
  private val tokenUrl = "https://auth.apis.do/user_management/authenticate"
  private val userUrl = "https://apis.do/me"

  /**
   * Initiates device authorization.
   */
  def authorize(scope: String): Try[DeviceAuthResponse] = {
    val body = Json.obj(
      "client_id" -> Json.fromString(clientId),
      "scope" -> Json.fromString(scope)
    )

    val request = basicRequest
      .post(uri"$authUrl")
      .contentType("application/json")
      .body(body.noSpaces)

    Try {
      val response = request.send(backend)
      response.body match {
        case Right(json) =>
          parse(json).flatMap { j =>
            val cursor = j.hcursor
            for {
              deviceCode <- cursor.get[String]("device_code")
              userCode <- cursor.get[String]("user_code")
              verificationUri <- cursor.get[String]("verification_uri")
              expiresIn <- cursor.getOrElse[Int]("expires_in")(900)
              interval <- cursor.getOrElse[Int]("interval")(5)
            } yield DeviceAuthResponse(deviceCode, userCode, verificationUri, expiresIn, interval)
          }.toTry.get
        case Left(error) => throw new Exception(s"Authorization failed: $error")
      }
    }
  }

  /**
   * Polls for tokens until user authorizes or timeout.
   */
  def pollForTokens(deviceCode: String, interval: Int, expiresIn: Int): Try[TokenResponse] = {
    val deadline = System.currentTimeMillis() + (expiresIn * 1000L)
    doPoll(deviceCode, interval, deadline)
  }

  private def doPoll(deviceCode: String, interval: Int, deadline: Long): Try[TokenResponse] = {
    if (System.currentTimeMillis() >= deadline) {
      Failure(new Exception("Authorization expired"))
    } else {
      val body = Json.obj(
        "client_id" -> Json.fromString(clientId),
        "device_code" -> Json.fromString(deviceCode),
        "grant_type" -> Json.fromString("urn:ietf:params:oauth:grant-type:device_code")
      )

      val request = basicRequest
        .post(uri"$tokenUrl")
        .contentType("application/json")
        .body(body.noSpaces)

      val response = request.send(backend)
      response.body match {
        case Right(json) =>
          parse(json).flatMap { j =>
            val cursor = j.hcursor
            cursor.get[String]("error") match {
              case Right("authorization_pending") =>
                Thread.sleep(interval * 1000L)
                doPoll(deviceCode, interval, deadline).toEither
              case Right("slow_down") =>
                Thread.sleep((interval + 5) * 1000L)
                doPoll(deviceCode, interval + 5, deadline).toEither
              case Right(error) =>
                Left(DecodingFailure(s"Error: $error", Nil))
              case Left(_) =>
                for {
                  accessToken <- cursor.get[String]("access_token")
                  refreshToken = cursor.get[String]("refresh_token").toOption
                  tokenType <- cursor.getOrElse[String]("token_type")("Bearer")
                  expiresIn = cursor.get[Int]("expires_in").toOption
                } yield TokenResponse(accessToken, refreshToken, tokenType, expiresIn)
            }
          }.toTry
        case Left(error) => Failure(new Exception(s"Token request failed: $error"))
      }
    }
  }

  /**
   * Gets user info with access token.
   */
  def getUserInfo(accessToken: String): Try[Json] = {
    val request = basicRequest
      .get(uri"$userUrl")
      .header("Authorization", s"Bearer $accessToken")

    Try {
      val response = request.send(backend)
      response.body match {
        case Right(json) => parse(json).toTry.get
        case Left(error) => throw new Exception(s"Failed to get user info: $error")
      }
    }
  }
}

File: packages/swift/oauth/Sources/OAuthDo/DeviceFlow.swift
import Foundation

/// Device flow authorization for OAuth.
/// Implements the device authorization grant flow.
public class DeviceFlow {
    private static let authURL = "https://auth.apis.do/user_management/authorize_device"
    private static let tokenURL = "https://auth.apis.do/user_management/authenticate"
    public static let defaultClientId = "client_01JQYTRXK9ZPD8JPJTKDCRB656"

    private let clientId: String
    private let session: URLSession

    /// Device authorization response.
    public struct DeviceAuthResponse {
        public let deviceCode: String
        public let userCode: String
        public let verificationUri: String
        public let verificationUriComplete: String?
        public let expiresIn: Int
        public let interval: Int

        public var description: String {
            "Visit \(verificationUri) and enter code: \(userCode)"
        }
    }

    /// Token response from authentication.
    public struct TokenResponse {
        public let accessToken: String?
        public let refreshToken: String?
        public let tokenType: String
        public let expiresIn: Int
        public let error: String?
        public let errorDescription: String?

        public var isError: Bool { error != nil }
        public var isPending: Bool { error == "authorization_pending" || error == "slow_down" }
    }

    /// Device flow errors.
    public enum DeviceFlowError: Error, LocalizedError {
        case authorizationFailed(String)
        case timeout
        case invalidResponse

        public var errorDescription: String? {
            switch self {
            case .authorizationFailed(let message): return "Authorization failed: \(message)"
            case .timeout: return "Authorization timed out"
            case .invalidResponse: return "Invalid response from server"
            }
        }
    }

    /// Creates a DeviceFlow with the default client ID.
    public init() {
        self.clientId = Self.defaultClientId
        self.session = URLSession.shared
    }

    /// Creates a DeviceFlow with a custom client ID.
    /// - Parameter clientId: OAuth client ID
    public init(clientId: String) {
        self.clientId = clientId
        self.session = URLSession.shared
    }

    /// Initiates device authorization.
    /// - Returns: Device authorization response
    /// - Throws: DeviceFlowError if authorization fails
    public func authorize() async throws -> DeviceAuthResponse {
        var request = URLRequest(url: URL(string: Self.authURL)!)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = "client_id=\(clientId)".data(using: .utf8)

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw DeviceFlowError.authorizationFailed(String(data: data, encoding: .utf8) ?? "Unknown error")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DeviceFlowError.invalidResponse
        }

        guard let deviceCode = json["device_code"] as? String,
              let userCode = json["user_code"] as? String,
              let verificationUri = json["verification_uri"] as? String else {
            throw DeviceFlowError.invalidResponse
        }

        return DeviceAuthResponse(
            deviceCode: deviceCode,
            userCode: userCode,
            verificationUri: verificationUri,
            verificationUriComplete: json["verification_uri_complete"] as? String,
            expiresIn: json["expires_in"] as? Int ?? 900,
            interval: json["interval"] as? Int ?? 5
        )
    }

    /// Polls for token after user authorization.
    /// - Parameters:
    ///   - deviceCode: Device code from authorization response
    ///   - interval: Polling interval in seconds
    ///   - timeout: Maximum time to wait in seconds
    ///   - onPoll: Callback for each poll attempt (optional)
    /// - Returns: Token response
    /// - Throws: DeviceFlowError if polling fails
    public func pollForToken(
        deviceCode: String,
        interval: Int = 5,
        timeout: Int = 900,
        onPoll: ((Int) -> Void)? = nil
    ) async throws -> TokenResponse {
        let startTime = Date()
        var pollInterval = interval
        var attempts = 0

        while Date().timeIntervalSince(startTime) < Double(timeout) {
            attempts += 1
            onPoll?(attempts)

            let body = "client_id=\(clientId)&device_code=\(deviceCode)&grant_type=urn:ietf:params:oauth:grant-type:device_code"

            var request = URLRequest(url: URL(string: Self.tokenURL)!)
            request.httpMethod = "POST"
            request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
            request.httpBody = body.data(using: .utf8)

            let (data, _) = try await session.data(for: request)

            guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw DeviceFlowError.invalidResponse
            }

            if let error = json["error"] as? String {
                if error == "slow_down" {
                    pollInterval += 5
                }

                if error != "authorization_pending" && error != "slow_down" {
                    return TokenResponse(
                        accessToken: nil,
                        refreshToken: nil,
                        tokenType: "Bearer",
                        expiresIn: 0,
                        error: error,
                        errorDescription: json["error_description"] as? String
                    )
                }
            } else if let accessToken = json["access_token"] as? String {
                return TokenResponse(
                    accessToken: accessToken,
                    refreshToken: json["refresh_token"] as? String,
                    tokenType: json["token_type"] as? String ?? "Bearer",
                    expiresIn: json["expires_in"] as? Int ?? 3600,
                    error: nil,
                    errorDescription: nil
                )
            }

            try await Task.sleep(nanoseconds: UInt64(pollInterval) * 1_000_000_000)
        }

        throw DeviceFlowError.timeout
    }

    /// Gets the client ID.
    /// - Returns: OAuth client ID
    public func getClientId() -> String {
        return clientId
    }
}
